src/isomorphic/classic/element/ReactElement.js

```diff
@@ -123,8 +126,10 @@ ReactElement.createElement = function(type, config, children) {
   var source = null;
 
   if (config != null) {
-    ref = config.ref === undefined ? null : config.ref;
-    key = config.key === undefined ? null : '' + config.key;
+    ref = !config.hasOwnProperty('ref') ||
+      Object.getOwnPropertyDescriptor(config, 'ref').get ? null : config.ref;
```

Comment:
The behaviour change can be handled with another condition (check for `undefined`), making the checks more ugly.